core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java

```diff
@@ -19,23 +19,665 @@
+    }
+
+    public void testStartEndObject() throws IOException {
+        expectUnclosedException(() -> builder().startObject().bytes());
+        expectUnclosedException(() -> builder().startObject().close());
+        expectUnclosedException(() -> builder().startObject().string());
+
+        expectObjectException(() -> builder().endObject().bytes());
+        expectObjectException(() -> builder().endObject().close());
+        expectObjectException(() -> builder().endObject().string());
+
+        expectValueException(() -> builder().startObject("foo").endObject());
+        expectNonNullFieldException(() -> builder().startObject().startObject(null).endObject().endObject());
+
+        assertResult(() -> builder().startObject().endObject(), equalTo("{}"));
+        assertResult(() -> builder().startObject().startObject("foo").endObject().endObject(), equalToJSon("{'foo':{}}"));
+        assertResult(() -> builder().startObject().startObject("foo").startObject("bar").endObject().endObject().endObject(),
+                equalToJSon("{'foo':{'bar':{}}}"));
+    }
+
+    public void testStartEndArray() throws IOException {
+        expectUnclosedException(() -> builder().startArray().bytes());
+        expectUnclosedException(() -> builder().startArray().close());
+        expectUnclosedException(() -> builder().startArray().string());
+
+        expectArrayException(() -> builder().endArray().bytes());
+        expectArrayException(() -> builder().endArray().close());
+        expectArrayException(() -> builder().endArray().string());
+
+        expectValueException(() -> builder().startArray("foo").endObject());
+        expectFieldException(() -> builder().startObject().startArray().endArray().endObject());
+        expectNonNullFieldException(() -> builder().startObject().startArray(null).endArray().endObject());
+
+        assertResult(() -> builder().startObject().startArray("foo").endArray().endObject(), equalToJSon("{'foo':[]}"));
+    }
+
+    public void testField() throws IOException {
+        expectValueException(() -> builder().field("foo").bytes());
+        expectNonNullFieldException(() -> builder().field(null).bytes());
+        expectUnclosedException(() -> builder().startObject().field("foo").bytes());
+
+        assertResult(() -> builder().startObject().field("foo").value("bar").endObject(), equalToJSon("{'foo':'bar'}"));
+    }
+
+    public void testNullField() throws IOException {
+        expectValueException(() -> builder().nullField("foo").bytes());
+        expectNonNullFieldException(() -> builder().nullField(null).bytes());
+        expectUnclosedException(() -> builder().startObject().nullField("foo").bytes());
+
+        assertResult(() -> builder().startObject().nullField("foo").endObject(), equalToJSon("{'foo':null}"));
+    }
+
+    public void testNullValue() throws IOException {
+        assertResult(() -> builder().startObject().field("foo").nullValue().endObject(), equalToJSon("{'foo':null}"));
```

Comment:
Can you make it `equalToJson`? I like the rules [here](https://google.github.io/styleguide/javaguide.html#s5.3-camel-case) because it makes it easy to tell where the words start and end.